Support TakeOrderedAndProject (#1579)
NVIDIA/spark-rapids: docs/FAQ.md

query. This is why `spark.rapids.sql.enabled` is still respected when running, even if explain shows
stale results.

### Why does the plan for the GPU query look different from the CPU query?

Typically, there is a one to one mapping between CPU stages in a plan and GPU stages.  There are a
few places where this is not the case.

> * `WholeStageCodeGen` - The GPU plan typically does not do code generation, and does not support generating code for an

Reviewer: I'd like to keep the line lengths to 100 characters for ease of editing, where possible.

> * `WholeStageCodeGen` - The GPU plan typically does not do code generation, and does not support generating code for an

Author: I can see us trying to limit ourselves to 100 characters most of the time, I have real problems from a change management perspective when we try to make all of the lines as close to 100 characters as possible, like most of this file is.  When you make a single small change to a line that forces it over 100 characters instantly you feel like you need to reformat the rest of the paragraph to make all of the lines as close to 100 characters as possible.  This means git sees a much bigger change and trying to merge conflicting changes becomes a real pain.  Especially because all of those changes have no impact to what the final rendered result looks like.

entire stage in the plan. Code generation is typically used to reduce the cost of processing data

> one row at a time. The GPU plan processes the data in a columnar format, so the costs are different.

Reviewer 2: Would be good to elaborate here, e.g.: mentioning GPU data is processed in batches of rows, and batches often contain many thousands of rows.

> * ColumnarToRow and RowToColumnar Transitions - The CPU version of Spark plans typically process data in a row based format. The main exception to

Reviewer: Should ColumnarToRow and RowToColumnar be backquoted, like, for example, `WholeStageCodeGen` is in the previous section?  

```suggestion
* `ColumnarToRow` and `RowToColumnar` transitions - The CPU version of Spark plans typically process data in a row based format. The main exception to
```

> this is reading some kinds of columnar data, like parquet. When transitioning between the CPU and
> the GPU require transitions between row and columnar formatted data.

Reviewer 2: ```suggestion
this is reading some kinds of columnar data, like Parquet. Transitioning between the CPU and
the GPU requires transitioning between row and columnar formatted data.
```

> * `GpuCoalesceBatches` and `GpuShuffleCoalesce` - Processing data on the GPU often scales sublinerarly. That means doubling the data does takes less

Reviewer 2: ```suggestion
* `GpuCoalesceBatches` and `GpuShuffleCoalesce` - Processing data on the GPU scales sublinearly. That means doubling the data often takes less
```

Reviewer: Would it be helpful to add a sentence to indicate why processing on the GPU scales sublinearly?

than half the time. Because of this we want to process larger batches of data when possible. These

> operators will try to build larger batches of data to process.

Reviewer 2: ```suggestion
operators collect smaller batches of data into fewer, larger batches to process.
```

* `SortMergeJoin` - The RAPIDS accelerator does not support sort merge joins yet. For now we
translate sort merge joins into shuffled hash joins. Because of this there are times when

> Sorts may be removed or other sorts added to meet the ordering requirements of the query.

Reviewer 2: ```suggestion
sorts may be removed or other sorts added to meet the ordering requirements of the query.
```

* `TakeOrderedAndProject` - The `TakeOrderedAndProject` operator will take the top N entries in each task,
shuffle the results to a single executor and then take the top N results from that. The GPU

> plan often has more metrics than the CPU versions do, and when we tried to combine all of these
> operations into a single stage the metrics where confusing to understand what was happening. Instead
> we split the single stage up into multiple smaller parts so the metrics are clearer.

Reviewer: ```suggestion
plan often has more metrics than the CPU versions do, and when we tried to combine all of these
operations into a single stage the metrics were confusing to understand. Instead we split the single 
stage into multiple smaller parts so the metrics are clearer.
```
